src/main/groovy/com/lesfurets/jenkins/unit/DeclarativePipelineHarness.groovy

```diff
@@ -0,0 +1,157 @@
+        }
+        callStack.run(idx, doClear)
+        return this
+    }
+
+    public <T> DeclarativePipelineHarness runParameterizedClosureStep(String name, T value, Boolean doClear = true) {
+        ParameterizedCallStack<Object> callStack = parameterizedStepCallStacks[name]
+        if(callStack== null) {'pipeline'
+            throw new RuntimeException("No such call stack ${name}")
+        }
+        this.lastClosureStepParameter = callStack.run(value, doClear)
+        return this
+    }
+
+    DeclarativePipelineHarness runParameterizedClosureStepByIndex(String name, Integer idx = 0, Boolean doClear = true) {
+        ParameterizedCallStack<Object> callStack = parameterizedStepCallStacks[name]
+        if(callStack== null) {
+            throw new RuntimeException("No such call stack ${name}")
+        }
+        this.lastClosureStepParameter = callStack.run(idx, doClear)
+        return this
+    }
+
+    DeclarativePipelineHarness runStage(String name) {
+        return this
+                .runClosureStep('pipeline')
+                .runClosureStep('stages')
+                .runParameterizedClosureStep('stage', name)
+    }
+
+    DeclarativePipelineHarness runParallelSubStage(String parentName, String childName) {
+        return this
+                .runClosureStep('pipeline')
+                .runClosureStep('stages')
+                .runParameterizedClosureStep('stage', parentName)
+                .runClosureStep('parallel')
+                .runParameterizedClosureStep('stage', childName)
+    }
+
+    Object getLastClosureStepParameterValue() {
+        return lastClosureStepParameter
+    }
+
+    void registerClosureStep(String name) {
+        stepCallStacks.put(name, new CallStack(name, base.helper))
+    }
+
+    public <T> void registerParameterizedClosureStep(String name, Class<T> clazz) {
+        parameterizedStepCallStacks.put(name, new ParameterizedCallStack<T>(name, base.helper, clazz))
+    }
+
+    void clearAllCalls() {
+        base.helper.clearCallStack()
+        stepCallStacks.values().forEach { it.clearCalls() }
+        parameterizedStepCallStacks.values().forEach { it.clearCalls() }
+    }
+
+
+    class CallStack {
+        String name
+        PipelineTestHelper helper
+
+        List<Closure> calls = []
+
+        CallStack(name, helper) {
+            this.name = name
+            this.helper = helper
+
+            this.helper.registerAllowedMethod(name, [Closure.class], { calls.add(it) })
+        }
+
+        void clearCalls() {
+            calls.clear()
+        }
+
+        void run(Integer idx = 0, doClear = true) {
+            if(! (idx < calls.size())) {
+                throw new RuntimeException("not enough calls in stack for ${name}, expected idx: ${idx}, size: ${idx}")
+            }
+            def c = calls[idx]
+            if(doClear) { clearAllCalls() }
+            c()
+        }
+    }
+
+    class ParameterizedCallStack<T> {
```

Comment:
`ParameterizedCallStack` and `CallStack` looks almost the same, 
 i'm just wondering if these classes could be coupled by inheritance e.g. 
```
class ParameterizedCallStack extends CallStack
```

Reply:
or maybe replace it with just one class, 
As result you can reduce the logic splits you need to handle parameterized or non-parameterized closures.